api/v1/contexts/PKPContextHandler.inc.php

```diff
@@ -0,0 +1,490 @@
+
+		return parent::authorize($request, $args, $roleAssignments);
+	}
+
+	/**
+	 * Get a collection of contexts
+	 *
+	 * @param $slimRequest Request Slim request object
+	 * @param $response Response object
+	 * @param $args array arguments
+	 * @return Response
+	 */
+	public function getContexts($slimRequest, $response, $args) {
+		$request = $this->getRequest();
+		$contextService = ServicesContainer::instance()->get('context');
+
+		$defaultParams = array(
+			'count' => 20,
+			'offset' => 0,
+		);
+
+		$requestParams = array_merge($defaultParams, $slimRequest->getQueryParams());
+
+		$allowedParams = array();
+
+		// Process query params to format incoming data as needed
+		foreach ($requestParams as $param => $val) {
+			switch ($param) {
+				case 'isEnabled':
+					$allowedParams[$param] = (bool) $val;
+					break;
+
+				case 'searchPhrase':
+					$allowedParams[$param] = trim($val);
+					break;
+
+				case 'count':
+					$allowedParams[$param] = min(100, (int) $val);
+					break;
+
+				case 'offset':
+					$allowedParams[$param] = (int) $val;
+					break;
+			}
+		}
+
+		\HookRegistry::call('API::contexts::params', array(&$allowedParams, $slimRequest));
+
+		// Anyone not a site admin should not be able to access contexts that are
+		// not enabled
+		if (empty($allowedParams['isEnabled'])) {
+			$userRoles = $this->getAuthorizedContextObject(ASSOC_TYPE_USER_ROLES);
+			$canAccessDisabledContexts = !empty(array_intersect(array(ROLE_ID_SITE_ADMIN), $userRoles));
+			if (!$canAccessDisabledContexts) {
+				return $response->withStatus(403)->withJsonError('api.contexts.403.requestedDisabledContexts');
+			}
+		}
+
+		$items = array();
+		$contexts = $contextService->getContexts($allowedParams);
+		if (!empty($contexts)) {
+			$propertyArgs = array(
+				'request' => $request,
+				'slimRequest' => $slimRequest,
+			);
+			foreach ($contexts as $context) {
+				$items[] = $contextService->getSummaryProperties($context, $propertyArgs);
+			}
+		}
+
+		$data = array(
+			'itemsMax' => $contextService->getContextsMaxCount($allowedParams),
+			'items' => $items,
+		);
+
+		return $response->withJson($data, 200);
+	}
+
+	/**
+	 * Get a single context
+	 * @param $slimRequest Request Slim request object
+	 * @param $response Response object
+	 * @param array $args arguments
+	 *
+	 * @return Response
+	 */
+	public function getContext($slimRequest, $response, $args) {
+		$request = $this->getRequest();
+		$user = $request->getUser();
+
+		$contextService = ServicesContainer::instance()->get('context');
+		$context = $contextService->getContext((int) $args['contextId']);
+
+		if (!$context) {
+			return $response->withStatus(404)->withJsonError('api.contexts.404.contextNotFound');
+		}
+
+		// Don't allow to get one context from a different context's endpoint
+		if ($request->getContext() && $request->getContext()->getId() !== $context->getId()) {
+			return $response->withStatus(403)->withJsonError('api.contexts.403.contextsDidNotMatch');
```

Comment:
This check is implemented in a lot of different places, and besides the code duplication, it becomes really easy to overlook it somewhere. Nasty as it is in many ways, that's one of the benefits of the authorization toolset. Maybe Slim offers something we could use? Alternately, I've been wondering if we could look to Laravel for something, as I suspect we'll be using more of that framework (e.g. possibly more of its database layer).

Reply:
I'm struggling to think of a better approach here. Here are the things I'm thinking of:

1. Perform a general check in the `authorize` method. One way or another we would still need to know which endpoints require the check. So we'd either need to add it to the endpoint definition (where it is also easily overlooked and we can't return a meaningful error message) or we'd need to check for the `contextId` arg (which would mean the auth check would be dropped if someone decided to name a new endpoint's argument `context`).

2. Extract the three lines to a helper function. This strikes me as unnecessary since it is just three lines, repeated four times, and we would still have to return the response in the parent method. So the three lines would just become two:

```
$response = $this->checkContexts();
return $response;
```

3. Try to come up with a slim-style middleware that is generalisable to other objects, that is intelligent enough to determine from the object what context it belongs to and check this against the `Request` context. Something like this would be nice, if it was reliable. And it could be tied into the service classes, so that each service class had a `getObjectContext()` method. But I'm struggling to see how to do this without a lot of code and db reads that duplicate our existing policies.

My hesitation, too, is that the logic for each check is different, and in some cases may be inappropriate. For example, in the future we may want to show notifications, discussions, and submission updates from a "news feed" type display for editors and assistants, which would display activity across contexts. In such a case, we would want to bypass the middleware for such endpoints.

So I'm not sure how generalisable this problem is, and in such cases I usually prefer to have a little bit of code duplication that is right there in the request handler rather than something under-the-hood that escapes attention.

But I'm open to other ideas. This issue -- how the request context can be different from the object context -- is a persistent issue in the API that I think will continue to cause headaches. I just haven't been able to come up with a good way too tackle it.

Reply:
The way the authorization framework tackled this was to tie together the fetching of an object (typically from the DB) with its validation. The risk of introducing security flaws typically happens when an object is fetched but never validated.

I think at some point we're going to need to move our current policy framework over into something standard -- like https://laravel.com/docs/5.7/authorization -- and that may help solve the problem here as well. But that'll certainly overload this task. Mind filing it for later?

Reply:
@asmecher https://github.com/pkp/pkp-lib/issues/4238